### ISSUE TITLE
[[ -v ]] accepts an expression

### ISSUE THREAD

**Issue by andychu** [oilshell/oil#1532]
from e-mail

```
$ declare -A foo
$ foo['bar']=2
$ [[ -v foo['bar'] ]] && echo exists
exists
$

Behaviour in OSH:
$ declare -A foo
$ foo['bar']=2
$ [[ -v foo['bar'] ]] && echo exists
$ 
```

I would like to know if this blocks a real script from running

**Comment by Melkor333** [comment]
Yes it does kinda. I just wanted to check if an assoc array has a key :(
```bash
    if [[ -v "package_attrsets[$1]" ]]; then
        return
    fi
```
What would be your alternative option to test if a key is in an associative array in `osh`/`bash`?
The [stackoverflow answer](https://stackoverflow.com/questions/13219634/easiest-way-to-check-for-an-index-or-a-key-in-an-array) is IMO extremely ugly

**Comment by andychu** [comment]
Hm I'd be curious to see the script

I don't actually recommend using associative arrays in bash at all, because they have pretty weird behavior, and it has actually changed a fair bit from version to version

The `-v` thing has apparently changed behavior as well (help-bash@ has a thread about this right now)

Does `${assoc[$1]:-}` work for you?  If so that seems like a pretty good answer


In YSH it would just be `if ('key' in package_attrsets)`, as in python

**Comment by andychu** [comment]
If there is something that really needs this, we can consider it a feature request of course

**Comment by andychu** [comment]
Hm yes this should either be implemented or put on "Known differences"

I wasn't aware you had a use case for this

Can you show me the original and the rewrite?

**Comment by Melkor333** [comment]
> Can you show me the original and the rewrite?

I wrote the script from scratch and didn't commit the "bash-only" way. I wrote a first version, switched to OSH, fixed this (and some other issues) and commited it afterwards.

So having it in the known differences is enough for me I guess.

**Comment by andychu** [comment]
> The annoying thing is I can't find this behavior documented
> 
> e.g. here
> 
> https://www.gnu.org/software/bash/manual/bash.html#index-_005b_005b
> 
> or here
> 
> https://www.gnu.org/software/bash/manual/bash.html#index-test
> 
> or here
> 
> ```
> -v varname
> 
>     True if the shell variable varname is set (has been assigned a value).
> ```
> 
> and also apparently there was a change in bash 5.2 around `[[ -v a[@] ]]`

**Comment by Melkor333** [comment]
77 upvotes on stackoverflow: https://stackoverflow.com/questions/13219634/easiest-way-to-check-for-an-index-or-a-key-in-an-array

that's how I found it. And others have, too…
Compared to the accepted answer it *looks* far less like a hack. But I have to agree it's not obvious from the man page and if it's not even officially (clearly) documented then IMO it's more than fine to keep it as a known difference.

**Comment by bar-g** [comment]
When [eval_unsafe_arith](https://www.oilshell.org/release/0.19.0/doc/ref/chap-option.html#eval_unsafe_arith) is disabled, all of `[ -v`, `[[ v` and `${!...}`, etc. should become safe again though, is that correct?

**Comment by andychu** [comment]
Yes, that option globally disables the dynamic parsing and evaluation

It still does some array subscripting, but not the infinite recursion that bash does!

**Comment by andychu** [comment]
This issue is similar to #1780 

`${!indirect}` is basically the same mechanism as `-v` 

So we could implement `-v` technically, though it appears to be undocumented by bash

**Comment by bar-g** [comment]
When run `man bash` and seach for it by entering `/-v var`, the first hit is:
```

       -v varname
              True if the shell variable varname is set (has been assigned a value).
```

**Comment by andychu** [comment]
Right but it doesn't say that `a[0]` and `A[key]` are accepted

we already support `-v` for what's documented

**Comment by bar-g** [comment]
Oh, yes I missed that angle here. Maybe they feel it's covered by the notion of bash array cells being compatible with classic variables?
